DOP-2325 - Set up Realm and write responses

carooobregon opened the pull request:
### Stories/Links:

DOP-2325
DOP-2381

### Staging Links:

[Staging](https://docs-mongodbcom-staging.corp.mongodb.com/master/cloud-docs/carolinao/DOP-2325/mongodb-logs/)

### Notes:
Used the [realm-template-apps](https://github.com/mongodb-university/realm-template-apps/tree/main/generated/web-mql) to write the Provider and Context needed to set up Realm

rayangler commented on src/components/Widgets/QuizWidget/RealmApp.js:
```diff
@@ -0,0 +1,38 @@
+
+export function RealmAppProvider({ appId, children }) {
+  // Store Realm.App in React state. If appId changes, all children will rerender and use the new realmApp.
+  const [realmApp, setRealmApp] = React.useState(createRealmApp(appId));
```
Is `setRealmApp` supposed to be used anywhere?

carooobregon replied:
not right now! tbh i just thought every state had to have a `set` function, should i delete it?

rayangler replied:
I think using it with `useEffect` like how they do [in the template](https://github.com/mongodb-university/realm-template-apps/blob/main/generated/web-mql/src/components/RealmApp.js#L13-L16) wouldn't be such a bad idea. Usually if we have a state, the expectation is that we have a corresponding `setState` function to update it

rayangler commented on src/layouts/index.js:
```diff
@@ -79,9 +81,11 @@ const DefaultLayout = (props) => {
-        <ContentsProvider headingNodes={page?.options?.headings}>
-          <Template {...props}>{children}</Template>
-        </ContentsProvider>
+        <RealmAppProvider appId={appId}>
```
What's the purpose of placing the quiz widget's context provider this high up vs. just wrapping the `QuizWidget` with this? Just wondering in case there was a specific reason!

Also, if this `RealmAppProvider` is specifically for the `QuizWidget`, do we need to pass the `appId` as a prop to the provider component instead of just keeping it inside of the component?

carooobregon replied:
1. You're right! I just changed it so it's wrapped around `SubmitButton` since all the submit logic is inside there :-) 
2. In theory, `RealmAppProvider` can be reused in posteriority for other Realm apps we want to use in snooty, hence sending the `appId` as a prop, however as we're only using it for QuizWidget rn I could change it so that it gathers the `appId` from the `realm.json` as a constant inside of the component instead of sending it as a prop, what do you think is better?

rayangler replied:
Ah, okay, thanks for the explanation! I think being able to reuse it would be cool, especially for the upcoming Feedback Widget, so sending in the id as a prop sounds good to me

rayangler commented on tests/unit/QuizWidget.test.js:
```diff
@@ -8,10 +8,15 @@ import { completeQuiz, noQuestion } from './data/QuizWidget.test.json';
   useSiteMetadata: () => ({ siteUrl, project }),
 }));
 
+beforeAll(() => {});
```
Do we still need this here?

carooobregon replied:
nope thx thx

rayangler commented on src/components/Widgets/QuizWidget/RealmApp.js:
```diff
@@ -0,0 +1,38 @@
+    logIn();
+  }, [logIn]);
+
+  // Override the App's currentUser & logIn properties + include the app-level logout function
```
Not sure if I'm misunderstanding this comment but what exactly is this "logout function"?

carooobregon replied:
oh it carried over from the [realm-template-apps](https://github.com/mongodb-university/realm-template-apps/tree/main/generated/web-mql) comments bc they had a logout function but i just deleted it !

rayangler commented on tests/unit/QuizWidget.test.js:
```diff
@@ -21,3 +26,7 @@ it('doesnt render quiz widget without a specified question', () => {
 });
+
+afterAll(() => {
+  spy.mockRestore();
```
This is a nice touch

rayangler reviewed (commented):
This is very cool. Just a few questions for now!

mango-db commented on src/components/Widgets/QuizWidget/RealmFuncs.js:
```diff
@@ -0,0 +1,17 @@
+export function useRealmFuncs() {
+  const responseCollection = useCollection({
+    cluster: dataSourceName,
+    db: 'quiz_test',
```
Is this db supposed to be `quiz_test` in production?

carooobregon replied:
ohh very true! just added `dbName` as a parameter in `useRealmFuncs` to support different databases to write to and also  added `const dbName = snootyEnv == "production" ? "quiz_prod" : "quiz_dev";` to decide which one to send, it seems as `snootyEnv` from `useSiteMetadata()` doesn't catch staging as a snootyEnv so I just left `quiz_dev` and `quiz_prod`

rayangler reviewed (approved):
LGTM, all of my comments/concerns have been resolved. Thank you!

mango-db commented on src/components/Widgets/QuizWidget/QuizWidget.js:
```diff
@@ -9,6 +9,9 @@ import ComponentFactory from '../../ComponentFactory';
 import { theme } from '../../../theme/docsTheme';
 import { useSiteMetadata } from '../../../hooks/use-site-metadata';
 import { getPlaintext } from '../../../utils/get-plaintext';
+import { useRealmFuncs } from './RealmFuncs';
```
Nit: is there a more meaningful name we could give this function?

mango-db commented on src/components/Widgets/QuizWidget/QuizWidget.js:
```diff
@@ -78,23 +86,44 @@ const SubmitButton = ({ setIsSubmitted, selectedResponse, quizResponseObj }) =>
+      correctAnswerCount += 1;
+    }
+  });
+  return correctAnswerCount;
```
You may be able to one-liner this with something like:

`return choices.filter(c => !!c.options?.['is-true']).length === 1`

The filter can be used to reduce `choices` to an array of booleans. Then, you can take the array of that, and check that it is exactly equivalent to 1.

I haven't tested this so you may have to mess around with this ^^

mango-db reviewed (commented):
LGTM:

- [x] Correctly posts test data to quiz_dev in develop AND [staging](https://docs-mongodborg-staging.corp.mongodb.com/test-quiz-widget/cloud-docs/cgoecknerwald/realm-app/) mode, including correct and incorrect response
- [] Posts data to quiz_prod in production?
- [x] Functions on mobile

A thought: It would be interesting to see how this affects page load times + latency

mango-db commented on src/components/Widgets/QuizWidget/QuizWidget.js:
```diff
@@ -78,23 +86,44 @@ const SubmitButton = ({ setIsSubmitted, selectedResponse, quizResponseObj }) =>
 };
 
-const createQuizResponseObj = (question, quizId, selectedResponse, project) => {
+const createQuizResponseObj = (question, quizId, selectedResponse, project, quizDate) => {
```
Lol ^^ `:date:` field

rayangler commented on src/components/Widgets/QuizWidget/QuizWidget.js:
```diff
@@ -78,23 +86,44 @@ const SubmitButton = ({ setIsSubmitted, selectedResponse, quizResponseObj }) =>
+  };
+};
+
+const getCorrectAnswerCount = (choices) => {
```
I agree that this should be enforced in the parser but I also think it might be fine to leave this logic here for now as well. It seems like there's a [ticket](https://jira.mongodb.org/browse/DOP-2386) that handles this check in the parser.

As a quick compromise, I'd say we can leave a comment here to let others know parser-level validation will be added later and we can check afterwards if this frontend check is still necessary

mango-db replied:
Yeeee, I agree 🤠

carooobregon commented:
@mango-db thanks for all your comments!! responding here bc github won't let me reply to comments
**date issue**: good catch! added verification using regex
**What does the "-ftaro" in "quizwidget-ftaro" mean?**
its the unique realm ID for the quiz widget app that is generated by Realm
**re. useRealmFuncs**
so in theory realmFuncs can have insert functions, delete functions, etc, since it's essentially the middleware between a DB and the snooty app, not necessarily quizwidget specific so that’s why it’s pretty generic and most likely will be used in feedback too, that being said i changed some names to make it more generic and so it can be reused, lmk if thats okay

mango-db reviewed (approved):
💎 [ʕ•́ᴥ•̀ʔっ♡](https://docs-mongodbcom-staging.corp.mongodb.com/test-quiz-widget/cloud-docs/cgoecknerwald/realm-app/
) 💎
